Add an option for multi-instrument callbacks (#2363)
open-telemetry/opentelemetry-specification: specification/metrics/api.md

  rule](#instrument-unit).
* An optional `description`, following the [instrument description
  rule](#instrument-description).
* Zero or more `callback` functions, responsible for reporting
  [Measurement](#measurement) values of the created instrument.

The API MUST support creation of asynchronous instruments by passing
zero or more `callback` functions to be permanently registered to the
newly created instrument.

A Callback is the conceptual entity created each time a `callback`
function is registered through an OpenTelemetry API.

The API SHOULD support registration of `callback` functions associated with
asynchronous instruments after they are created.

Where the API supports registration of `callback` functions after
asynchronous instrumentation creation, the user MUST be able to undo
registration of the specific callback after its registration somehow.

> Every currently registered Callback associated with an instrument MUST

Reviewer: ```suggestion
Every currently registered `callback` function associated with an instrument MUST
```

> be evaluted exactly once during collection prior to reading data for

Reviewer: What if there is timeout? What if one `callback` function failed (ideally, they shouldn't fail)?

Author: These are good questions, but would you agree that these questions are present with or without an option for multi-instrument callbacks?

  rule](#instrument-unit).
* An optional `description`, following the [instrument description
  rule](#instrument-description).
* Zero or more `callback` functions, responsible for reporting
  [Measurement](#measurement) values of the created instrument.

The API MUST support creation of asynchronous instruments by passing
zero or more `callback` functions to be permanently registered to the
newly created instrument.

The API SHOULD support registration of `callback` functions associated with
asynchronous instruments after they are created.

Where the API supports registration of `callback` functions after
asynchronous instrumentation creation, it MUST return something (e.g.,

> a registration handle, receipt or token) to the user that supports

Reviewer 2: Isn't this too language specific? I imagine there can be other ways to do this unregistration that are more idiomatic for different languages than using a returned object.

Reviewer 3: I can't think of one, but I always support making the spec less language specific :), so here to suggest simply saying the user must be able to undo registration of the specific callback after its registration. This implies that some unique reference must be available -- since you can't just say 'deregister everything registered on this instrument' to achieve the same thing.

But also having a, "for example, this could be done by returning a registratio handle, receipt or token", is useful to have, instead of a `MUST`.

Author: Changed this to 
```
Where the API supports registration of `callback` functions after
asynchronous instrumentation creation, the user MUST be able to undo
registration of the specific callback after its registration by some means.
```

Author: BTW @ocelotl I can't think of an idiom that doesn't involve returning something to the user. :-)
(I'm thinking of pointers as things returned to the user, so if you're somehow creating a pointer to a function you've got a thing returned to you.)

undoing the effect of `callback` registation.

A Callback is the conceptual entity created each time a `callback`
function is registered through an OpenTelemetry API.  

Every currently registered Callback associated with an instrument MUST
be evaluted exactly once during collection prior to reading data for

> that instrument.

Reviewer 2: Some languages may not need to create anything when a callback function is registered, they will just call that function during collection.

Author: Surely the SDK is going to store something to refer to the function that will be called. That thing, I'm arguing, is the Callback entity.

[OpenTelemetry API](../overview.md#api) authors MAY decide what is the idiomatic
approach for capturing measurements from callback functions. Here are some examples:

* Return a list (or tuple, generator, enumerator, etc.) of individual
  `Measurement` values.
* Pass an *Observable Result* as a formal parameter of the callback,
  where `result.Observe()` captures individual `Measurement` values.

Callbacks registered at the time of instrument creation MUST apply to

> the single instruments which is under construction.

Reviewer 4: Idea to investigate with maintainers: Can we actually remove this limitation? If we extend the "Result" to have 2 methods/funcs one that accepts an instrument as initial argument one that it does not, and let the user call whichever API they need, and probably error somehow if they call the non-instrument version for a callback registered to multiple instruments?

I am not sure how to cc all maintainers to this comment, so I will add all some sigs here:

@open-telemetry/go-maintainers 
@open-telemetry/java-maintainers 
@open-telemetry/python-maintainers 
@open-telemetry/dotnet-maintainers 
@open-telemetry/cpp-maintainers 
@open-telemetry/erlang-maintainers 
@open-telemetry/javascript-maintainers

> the single instruments which is under construction.

Reviewer 4: Idea to investigate with maintainers: Can we actually remove this limitation? If we extend the "Result" to have 2 methods/funcs one that accepts an instrument as initial argument one that it does not, and let the user call whichever API they need, and probably error somehow if they call the non-instrument version for a callback registered to multiple instruments?

I am not sure how to cc all maintainers to this comment, so I will add all some sigs here:

@open-telemetry/go-maintainers 
@open-telemetry/java-maintainers 
@open-telemetry/python-maintainers 
@open-telemetry/dotnet-maintainers 
@open-telemetry/cpp-maintainers 
@open-telemetry/erlang-maintainers 
@open-telemetry/javascript-maintainers

Reviewer 4: After more discussions with @jmacd, we need maintainers of stable API to tell us/show a prototype how are they planning to extend existent stable API to support this. We need to make sure we don't make already stable API ugly because of this.

Reviewer 5: FWIW, this is already a part of the Go API.

Reviewer 3: @MrAlias the restriction is part of the Go API? I can't understand the need for the restriction. Is there one in Go?

Author: The use of a multi-instrument form as the primary form is what OTel-Go does. We do have only one form there, but to me having both forms would be a minor improvement.  I won't speak for the other languages, I'm just surprised at the reaction that having two forms of callback is somehow confusing. Each callback is going to be called once, regardless of the form.

Callbacks registered after the time of instrument creation MAY be
associated with multiple instruments.

Idiomatic APIs for multiple-instrument Callbacks MUST distinguish the
instrument associated with each observed `Measurement` value.

Multiple-instrument Callbacks MUST be associated at the time of
registration with a declared set of asynchronous instruments from the

> same `Meter` instance.  This REQUIREMENT that Instruments be

Reviewer 2: REQUIREMENT is not an RFC2119 keyword.
```suggestion
same `Meter` instance.  This requirement that Instruments be
```

instruments from a single registered
Callback](#asynchronous-instrument-api).  The API to register a new
Callback SHOULD accept:

- A `callback` function
- A list (or tuple, etc.) of Instruments used in the `callback` function.

It is RECOMMENDED that the API authors use one of the following forms
for the `callback` function:

* The list (or tuple, etc.) returned by the `callback` function
  contains `(Instrument, Measurement)` pairs.
* the Observable Result parameter receives an additional `(Instrument,
  Measurement)` pairs

This interface is typically a more performant way to report multiple
measurements when they are obtained through an expensive process, such
as reading `/proc` files or probing the garbage collection subsystem.

For example,

```Python
# Python
class Device:
    """A device with two instruments"""

    def __init__(self, meter, proeprty):
        self.property = property
        self.usage = meter.create_observable_counter(name="usage", description="count of items used")
        self.pressure = meter.create_observable_gauge(name="pressure", description="force per unit area")

        # Note the two associated instruments are passed to the callback.

>         meter.register_callback([self.usage, self.pressure], self.observe)

Reviewer 6: It seems to me that the callback registered with `meter.create_observable_counter` and the callback registered with `meter.register_callback` have different signatures. In that case, I'd find it is confusing to have them all named "callback"s.

Author: I have been trying to make a distinction between "callback functions" (code) and Callbacks (concept/entity). I 

I'm not sure if it's overly subjective or not, to dictate a single form of callback function here. I believe the concept/entity are identical, even if the form of the callback function is not.

I do know this: if forced to pick a single callback API, I would choose the multi-instrument one even if it means that single-instrument callback functions are slightly more difficult to use. The performance benefits outweigh the disadvantages. I consider this to be not a problem because generally amount of code devoted to asynchronous instruments is small (unlike synchronous instruments).

@bogdandrutu has given feedback on this "forced decision", I think his opinion is that if we accept a multi-instrument callback API as in the example here, we should find a delicate way to deprecate the first form. I don't feel so strongly. 🤷

Reviewer 6: I have no strong opinion regarding the two forms of callbacks, forced to choose one or not. I just find that we have those two forms of callback named in a way that is hard to interpret or to be learned.

Reviewer 5: I don't think an implementation would use both forms. Meaning there won't be a confusion when this is implemented.

Reviewer 6: > I don't think an implementation would use both forms. Meaning there won't be a confusion when this is implemented.

If so, I think the spec should clearly state that implementation must only pick one of the patterns, but not do both.

Author: I don't see the conceptual difficulty with having more than one form of callback. I think of callbacks as things that produce measurements, and producing measurements for a single instrument is simpler than producing measurements for more than one instrument. The user is going to know which form makes sense for a given scenario. I can see one library using both forms, but I can't see one instrument using both forms.

* A value
* [`Attributes`](../common/common.md#attributes)

### Multiple-instrument callbacks

> Where idiomatic, [the Metrics API supports an option to use multiple
> instruments from a single registered
> callback](#asynchronous-instrument).  The key requirement for the API
> is that the association between Callbacks and Instruments be explicit
> so that the SDK knows which callbacks to execute when observing
> specific instruments.

Reviewer 7: ```suggestion
The Metrics API, if idiomatic, MAY support a way to use multiple instruments
from a single registered callback.  
If this API exisists it MUST explictly associate Callbacks with the Instruments
used in the callback.  This is because <insert the reason behind the key reuirement>
```

Author: Done. The <insert reason> has been clarified in the new `#asynchronous-instrument-api` section above.

> Some suggestions for the API author:

Reviewer 7: ```suggestion
It is RECOMMENDED that the API authors use one of the following forms:
```

Author: Done.

* the Observable result used in multi-instrument callbacks receives an

>   additional instrument argument

Reviewer 7: I don't understand this bullet, but that might be a lack of familiarity with Java's implementation.

Author: The "observable result" pattern is not named well above, it's this pattern:

```
def pf_callback(result):
  result.Observe(8, ("pid", 0), ("bitness", 64))
```

In OTel-Go we've been using the context.Context as a substitute for this "result" object.

Author: I've bolded the definition of the observable result pattern where it first appears, to make it clearer at this point.

* asynchronous instruments have an `Observe()` method that can be used
  from appropriately registered callbacks.

This interface is typically a more performant way to report multiple
measurements when they are obtained through an expensive process, such
as reading `/proc` files or probing the garbage collection subsystem.

For example,

```Python
# Python
class Device:
    """A device with two instruments"""

    def __init__(self, meter, proeprty):
        self.property = property
        self.usage = meter.create_observable_counter(name="usage", description="count of items used")
        self.pressure = meter.create_observable_gauge(name="pressure", description="force per unit area")

        # Note the two associated instruments are passed to the callback.

>         meter.register_callback([self.usage, self.pressure], self.observe)

Reviewer 4: Q: What does it happen if the instruments are not created via this Meter?

Author: It would be an error at the time of registration and/or at the time of use. By the library guidelines, these would be handled and not interrupt program execution.

    def observe(self, result):
        usage, pressure = expensive_system_call()
        self.usage.observe(result, usage, {'property', self.property})

>         self.pressure.observe(result, pressure, {'property', self.property})

Reviewer 4: why change the order "Observe" API is on the instrument instead on the result. Should we be consistent?

If we do on the `Instrument` that is a bit strange to have that API working only for `batch` but not for `single`.

Author: > why change the order "Observe" API is on the instrument instead on the result. Should we be consistent?

I didn't change the order, this is just a different suggestion than the one you're thinking of.  Two python examples are given for the simple case, one where a list of measurements is returned, and one where the "result" object is used to "observe" values. It's this second example that is inconsistent with the new example, but they're not required to go together.

```
def pf_callback(result):
    # Note: in the real world these would be retrieved from the operating system
    result.Observe(8,        ("pid", 0),   ("bitness", 64))
```

I don't think of the result as the one "doing" the observation in any case, I'd prefer to see that example written as "result.Add(observed_value, attrs...)" instead of "result.Observe(value, attrs...)" (i.e., "add a value to the result being observed").

> If we do on the Instrument that is a bit strange to have that API working only for batch but not for single.

Nothing I've written would prevent an instrument from being used in multiple callbacks, both single and multi-instrument forms. The fact that the two APIs have different shape is because one has an implicit instrument and one does not. You can create one instrument, register simple callbacks to provide some observations, and register batch callbacks to provide other observations.

In fact, the reason we're emphasizing this for OTel-Go is that the multi-instrument API is better than the simple API, and if possible we think having only a multi-instrument API makes sense. In the spirit of being idiomatic for Go, if you want to have a simpler API than the one provided, you'll have to write a few extra lines of code.

Author: I've changed the order to match the observable result pattern in order, i.e., it's

```
  result.observe(instrument, value, attributes...)
```
